proof_of_concept/registry.py

```diff
@@ -5,194 +5,279 @@
 
-        Raises:
-            KeyError: If no store with the given name is currently
-                    registered.
+        Args:
+            site_name: Name of the site where the runner is located.
+            admin: The party administrating this runner.
+            runner: The runner to register.
         """
-        return self._stores[name]
+        for i in self._store.objects():
+            if isinstance(i, RunnerDescription) and i.runner == runner:
+                raise RuntimeError(
+                        'There is already a runner called {}'.format(
+                            runner.name))
 
-    def get_store_admin(self, name: str) -> str:
-        """Look up who administrates a given store.
+        site = self._get_site(site_name)
+        runner_desc = RunnerDescription(site, runner)
+        self._store.insert(runner_desc)
+
+    def register_store(self, site_name: str, store: IAssetStore) -> None:
+        """Register an AssetStore with the Registry.
 
         Args:
-            name: The name of the store to look up.
+            site_name: The site this store is located at.
+            store: The data store to register.
+        """
+        for o in self._store.objects():
+            if isinstance(o, AssetStoreDescription) and o.store == store:
+                raise RuntimeError(
+                        'There is already a store called {}'.format(
+                            store.name))
 
-        Return:
-            The name of the party administrating it.
+        site = self._get_site(site_name)
+        store_desc = AssetStoreDescription(site, store)
+        self._store.insert(store_desc)
 
-        Raises:
-            KeyError: If no runner with the given name is registered.
+    def register_policy_server(
+            self, site_name: str, namespace_name: str, server: IPolicyServer
+            ) -> None:
+        """Register a PolicyServer with the registry.
+
+        Args:
+            site_name: Site at which this server is located.
+            namespace_name: The namespace this server serves policies
+                    for.
+            server: The data store to register.
         """
-        return self._store_admins[name]
+        for o in self._store.objects():
+            if isinstance(o, PolicyServerDescription) and o.server == server:
+                raise RuntimeError('The server is already registered')
 
-    def list_policy_servers(self) -> List[Tuple[str, IPolicyServer]]:
-        """List all known policy servers.
+        site = self._get_site(site_name)
 
-        Return:
-            A list of all registered policy servers and their
-                    namespaces.
+        for o in self._store.objects():
+            if isinstance(o, NamespaceDescription):
+                if o.name == namespace_name:
+                    namespace = o
+                    break
+        else:
+            raise RuntimeError('Namespace {} not found'.format(namespace_name))
```

Comment:
It would be clearer to me to replace this with a method `_get_object`, similar to `_object_exists` (see my comment on line 194). With this you could also simplify a lot of code in the ddm client it seems.

Reply:
Maybe in the `CanonicalStore` class

Reply:
I've added this now, but locally. I think it'd be good to refactor this into `CanonicalStore`, or into `Replica`, or maybe both. But that's probably better as as separate issue?

Reply:
Check, for now this looks good!